src/charm.py

```diff
@@ -6,36 +6,57 @@
 """Charm Jenkins."""
 
 import logging
-from typing import TYPE_CHECKING, Any, Dict, cast
+import typing
 
-from ops.charm import CharmBase, PebbleReadyEvent
+from ops.charm import ActionEvent, CharmBase, PebbleReadyEvent, RelationJoinedEvent
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus
 from ops.pebble import Layer
 
-from jenkins import JENKINS_WEB_URL, calculate_env, unlock_jenkins, wait_jenkins_ready
-from types_ import JenkinsEnvironmentMap
+import jenkins
+import state
 
-if TYPE_CHECKING:
+if typing.TYPE_CHECKING:
     from ops.pebble import LayerDict  # pragma: no cover
 
 
 logger = logging.getLogger(__name__)
 
 
+class AgentRelationData(typing.TypedDict):
```

Comment:
Should this be in an `Agent` module?

Reply:
Refactored to agent observer module! Thank you!